parsers/UY.py

```diff
@@ -1,163 +1,224 @@
 #!/usr/bin/python3
 
-import re
-from datetime import datetime
+import io
+from datetime import datetime, timedelta
 from logging import Logger, getLogger
 from typing import Optional
 
 import arrow
-import dateutil
+import pandas as pd
+import pytz
 
 # BeautifulSoup is used to parse HTML to get information
 from bs4 import BeautifulSoup
 from requests import Session
 
-tz = "America/Montevideo"
+from electricitymap.contrib.config.constants import PRODUCTION_MODES
+from parsers.lib.config import refetch_frequency
+from parsers.lib.exceptions import ParserException
 
-MAP_GENERATION = {
-    "Hidráulica": "hydro",
+UY_TZ = "America/Montevideo"
+
+ADME_URL = "https://pronos.adme.com.uy/gpf.php?fecha_ini="
+PRODUCTION_MODE_MAPPING = {
+    "Salto Grande": "hydro",
+    "Bonete": "hydro",
+    "Baygorria": "hydro",
+    "Palmar": "hydro",
     "Eólica": "wind",
-    "Fotovoltaica": "solar",
-    "Biomasa": "biomass",
+    "Solar": "solar",
     "Térmica": "oil",
+    "Biomasa": "biomass",
+}
+EXCHANGES_MAPPING = {
+    "Exp_Intercon_ARG": "AR->UY",
+    "Exp_Intercon_BR_MELO": "BR-S->UY",
+    "Exp_Intercon_BR_RIVERA": "BR-S->UY",
+    "Imp_Intercon_AG_Imp": "AR->UY",
+    "Imp_Intercon_BR_MELO": "BR-S->UY",
+    "Imp_Intercon_BR_RIVERA": "BR-S->UY",
 }
-INV_MAP_GENERATION = dict([(v, k) for (k, v) in MAP_GENERATION.items()])
-
-SALTO_GRANDE_URL = "http://www.cammesa.com/uflujpot.nsf/FlujoW?OpenAgent&Tensiones y Flujos de Potencia&"
-
-
-def get_salto_grande(session: Optional[Session]) -> float:
-    """Finds the current generation from the Salto Grande Dam that is allocated to Uruguay."""
-
-    current_time = arrow.now("UTC-3")
-    if current_time.minute < 30:
-        # Data for current hour seems to be available after 30mins.
-        current_time = current_time.shift(hours=-1)
-    lookup_time = current_time.floor("hour").format("DD/MM/YYYY HH:mm")
-
-    s = session or Session()
-    url = SALTO_GRANDE_URL + lookup_time
-    response = s.get(url)
-    soup = BeautifulSoup(response.text, "html.parser")
-
-    tie = soup.find("div", style="position:absolute; top:143; left:597")
-    generation = float(tie.text)
-
-    return generation
-
 
-def parse_page(session: Optional[Session]):
-    r = session or Session()
-    url = "https://apps.ute.com.uy/SgePublico/ConsPotenciaGeneracionArbolXFuente.aspx"
-    response = r.get(url)
-    soup = BeautifulSoup(response.text, "html.parser")
 
-    datefield = soup.find(
-        "span", attrs={"id": "ctl00_ContentPlaceHolder1_lblUltFecScada"}
+def get_adme_url(target_datetime: datetime, session: Optional[Session] = None) -> str:
```

Comment:
```suggestion
def get_adme_url(target_datetime: datetime, session: Session) -> str:
```
Since we are calling this function only from functions that handle the missing session case on their own, I would rather have this function just require a session.
This would add some clarity as the missing session problem would be treated all at the top and everything down just requires the session argument.